aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -403,6 +403,119 @@ class TypedOperatorHandle<Return (Args...)> final : public OperatorHandle {
 
 namespace detail {
 template<class... Args> inline void unused_arg_(const Args&...) {}
+
+// ReturnValue is intended to capture return values from Dispatcher unboxed
+// kernel calls.
+// A record function may requests to get outputs from the kernel calls. For
+// boxed kernels, it's straightforward, the returned values are in the stack
+// object. The stack can be passed to record functions. For unboxed kernels, we
+// need to handle different kinds of return values, cache them temporarily, then
+// release the values for the actual function call return.
+template <typename Return>
+struct ReturnValue {
+  Return val;
+  template <typename F, typename... Args>
+  ReturnValue(
+      F&& kernel,
+      const TypedOperatorHandle<Return(Args...)>& op,
+      DispatchKeySet& dispatchKeySet,
+      Args&&... args)
+      : val{kernel.template call<Return, Args...>(
+            op,
+            dispatchKeySet,
+            std::forward<Args>(args)...)} {}
+  // Wrap the return values in a vector of IValues. We also specialize for
+  // containers below, so they get
+  inline std::vector<IValue> getOutputs() {
+    std::vector<IValue> outputs;
+    outputs.emplace_back(IValue(val));
+    return outputs;
+  }
+  // Since we are returning the value, we don't expect the value to be used
+  // afterward. Copy elision and RVO do not apply to class data members. Using
+  // move semantic to avoid copies when possible.
+  inline Return release() {
+    return std::move(val);
```

Comment:
since you're moving from the state, this should be an `&&` overload and only callable on rvalue references, i.e.

```
Return release() && {
    return std::move(val);
}
```

Reply:
We do actually use `ReturnValue` object as a lvalue. `release()` follows a similar semantic as `std::unique_ptr::release()`.

Reply:
Thanks, I have made rvalue reference required for this method to enforce correctness.